### ISSUE TITLE
Develop local Python package not working since setuptools upgrade

### ISSUE THREAD

**Issue by cript0nauta** [NixOS/nixpkgs#108525]
**Describe the bug**

When following the steps in "15.19.1.3. Develop local package" section of the manual in nixpkgs-unstable, running a Python executable included in the package raises a `PackageNotFoundError` exception. It works ok in 20.09

**To Reproduce**
1. `git clone https://github.com/pypa/sampleproject`
2. `cd sampleproject`
2. Create a `shell.nix` file with the following contents:
```nix
with import <nixpkgs> {};
with python38Packages;

buildPythonPackage rec {
  name = "sampleproject";
  src = ./.;
  propagatedBuildInputs = [ peppercorn ];
}
```
3. Running `nix-shell --command 'sample'` will fail with the following traceback:
```
Traceback (most recent call last):
  File "/run/user/1000/tmp.xSLSr8Vrbm/bin/sample", line 33, in <module>
    sys.exit(load_entry_point('sampleproject', 'console_scripts', 'sample')())
  File "/run/user/1000/tmp.xSLSr8Vrbm/bin/sample", line 22, in importlib_load_entry_point
    for entry_point in distribution(dist_name).entry_points
  File "/nix/store/jdp800qwb3v330ggl6iq722ww96q2444-python3-3.8.6/lib/python3.8/importlib/metadata.py", line 504, in distribution
    return Distribution.from_name(distribution_name)
  File "/nix/store/jdp800qwb3v330ggl6iq722ww96q2444-python3-3.8.6/lib/python3.8/importlib/metadata.py", line 177, in from_name
    raise PackageNotFoundError(name)
importlib.metadata.PackageNotFoundError: sampleproject
```

**Expected behavior**
Running `nix-shell --command 'sample'` should output `Call your main application code here` with no exceptions.

**Additional context**
Commit ea0bf3c70a37d0a9a43019dd1c42037e03201dae introduced this regression. Something in the newer version of setuptools is causing the problem. I suspect it is related to https://github.com/pypa/setuptools/issues/2165 since it is a breaking change that already caused other problems while using Python projects with Nix.

**Notify maintainers**

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.81, NixOS, 21.03pre257780.e9158eca70a (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.9`
 - channels(criptonauta): `"nixpkgs-21.03pre259075.9cb4d2f4903"`
 - channels(root): `"nixos-21.03pre257780.e9158eca70a"`
 - nixpkgs: `/home/criptonauta/.nix-defexpr/channels/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
- python3Packages.setuptools
# a list of nixos modules affected by the problem
module:
```

**Comment by shuhaowu** [comment]
Are there any workarounds for this?

**Comment by cript0nauta** [comment]
I'm still working on a fix for this issue. A workaround is to manually add a site.py file inside the temp environment, as setuptools did in older versions. 

I contacted the author of the setuptools change for help, in order to make a less hacky fix than this. Here is the discussion: https://github.com/pypa/setuptools/issues/2165.

**Comment by FRidh** [comment]
Fixed with 7a65bb76f1db44f8af6e13d81d13f41d69fb1948.
